engine/src/main/java/io/camunda/zeebe/engine/state/appliers/EventAppliers.java

```diff
@@ -293,13 +294,22 @@ private void registerCommandDistributionAppliers(final MutableProcessingState st
   }
 
   private <I extends Intent> void register(final I intent, final TypedEventApplier<I, ?> applier) {
-    mapping.put(intent, applier);
+    register(intent, RecordMetadata.DEFAULT_RECORD_VERSION, applier);
+  }
+
+  private <I extends Intent> void register(
+      final I intent, final int version, final TypedEventApplier<I, ?> applier) {
+    mapping.put(new IntentAndVersion(intent, version), applier);
   }
 
   @Override
-  public void applyState(final long key, final Intent intent, final RecordValue value) {
+  public void applyState(
+      final long key, final Intent intent, final RecordValue value, final int recordVersion) {
     final var eventApplier =
-        mapping.getOrDefault(intent, UNIMPLEMENTED_EVENT_APPLIER.apply(intent));
+        mapping.getOrDefault(
+            new IntentAndVersion(intent, recordVersion), UNIMPLEMENTED_EVENT_APPLIER.apply(intent));
     eventApplier.applyState(key, value);
   }
+
+  record IntentAndVersion(Intent intent, int version) {}
```

Comment:
💭 Thinking ahead here. If we want to use the latest event applier by default we can just write a compare function for this record and search the map for the latest version of a specific intent.